File: hw1/solve.cc
#include <iostream>
#include <vector>
#include <algorithm>

//Пример кода бинарной кучи 
int main() {
    std::vector<int> heap = {10, 20, 15};
    std::make_heap(heap.begin(), heap.end());  // Создание кучи

    heap.push_back(30);
    std::push_heap(heap.begin(), heap.end());  // Вставка элемента

    std::pop_heap(heap.begin(), heap.end());   // Удаление корня (максимума)
    heap.pop_back();

    for(int i : heap) std::cout << i << " ";
    return 0;
}
//Hash таблицы 
#include <iostream>
#include <unordered_map>

int main() {
    std::unordered_map<std::string, int> hash_map;
    hash_map["one"] = 1;
    hash_map["two"] = 2;

    std::cout << hash_map["one"];
    return 0;
}
//Биномиальная куча (Binomial Heap)
#include <iostream>
using namespace std;

// Узел биномиального дерева
struct Node {
    int key;
    int degree;
    Node* parent;
    Node* child;
    Node* sibling;

    Node(int k) : key(k), degree(0), parent(nullptr), child(nullptr), sibling(nullptr) {}
};

// Слияние двух биномиальных деревьев одинаковой степени
Node* binomialLink(Node* y, Node* z) {
    y->parent = z;
    y->sibling = z->child;
    z->child = y;
    z->degree++;
    return z;
}

// Объединение двух списков корней биномиальных куч
Node* binomialHeapMerge(Node* h1, Node* h2) {
    if (!h1) return h2;
    if (!h2) return h1;

    Node* head = nullptr;
    Node* tail = nullptr;
    Node* a = h1;
    Node* b = h2;

    if (a->degree <= b->degree) {
        head = a;
        a = a->sibling;
    } else {
        head = b;
        b = b->sibling;
    }
    tail = head;

    while (a && b) {
        if (a->degree <= b->degree) {
            tail->sibling = a;
            a = a->sibling;
        } else {
            tail->sibling = b;
            b = b->sibling;
        }
        tail = tail->sibling;
    }
    tail->sibling = a ? a : b;
    return head;
}

// Функция вставки нового элемента в биномиальную кучу
Node* binomialHeapInsert(Node* h, int key) {
    Node* newNode = new Node(key);
    return binomialHeapUnion(h, newNode);
}

// Объединение двух биномиальных куч
Node* binomialHeapUnion(Node* h1, Node* h2) {
    Node* newHeap = binomialHeapMerge(h1, h2);
    if (!newHeap) return nullptr;

    Node* prev = nullptr;
    Node* curr = newHeap;
    Node* next = curr->sibling;

    while (next != nullptr) {
        if ( (curr->degree != next->degree) || 
             (next->sibling != nullptr && next->sibling->degree == curr->degree)) {
            prev = curr;
            curr = next;
        } else {
            if (curr->key <= next->key) {
                curr->sibling = next->sibling;
                binomialLink(next, curr);
            } else {
                if (prev == nullptr)
                    newHeap = next;
                else
                    prev->sibling = next;
                binomialLink(curr, next);
                curr = next;
            }
        }
        next = curr->sibling;
    }
    return newHeap;
}

// Пример использования
int main() {
    Node* heap = nullptr;
    heap = binomialHeapInsert(heap, 10);
    heap = binomialHeapInsert(heap, 20);
    heap = binomialHeapInsert(heap, 15);

    // Далее можно реализовать функции извлечения минимума, удаления и т.д.
    cout << "Биномиальная куча построена с элементами 10, 20, 15." << endl;

    return 0;
}

//Двоичная куча

#include <iostream>
#include <vector>
#include <algorithm>
int main() {
    std::vector<int> heap = {10, 20, 15};
    std::make_heap(heap.begin(), heap.end());
    heap.push_back(30);
    std::push_heap(heap.begin(), heap.end());
    std::pop_heap(heap.begin(), heap.end());
    heap.pop_back();
    for (int i : heap) std::cout << i << " ";
}

File: hw1/solve.java
//код бинарной кучи (макс-кучи) в C++:

import java.util.PriorityQueue;
import java.util.Collections;

public class MaxHeapExample {
    public static void main(String[] args) {
        PriorityQueue<Integer> maxHeap = new PriorityQueue<>(Collections.reverseOrder());
        maxHeap.add(10);
        maxHeap.add(20);
        maxHeap.add(15);

        maxHeap.poll(); // Удаление максимума

        for(int i : maxHeap)
            System.out.print(i + " ");
    }
}

// Двоичная куча (Binary Heap)

import java.util.PriorityQueue;
import java.util.Collections;

public class BinaryHeap {
    public static void main(String[] args) {
        PriorityQueue<Integer> maxHeap = new PriorityQueue<>(Collections.reverseOrder());
        maxHeap.add(10);
        maxHeap.add(20);
        maxHeap.add(15);

        maxHeap.poll(); // Удаление максимума

        for (int i : maxHeap)
            System.out.print(i + " ");
    }
}
// Биномиальная куча (Binomial Heap)

class BinomialNode:
    def __init__(self, key):
        self.key = key
        self.children = []
        self.degree = 0

class BinomialHeap:
    def __init__(self):
        self.trees = []

    def merge_trees(self, t1, t2):
        if t1.key > t2.key:
            t1, t2 = t2, t1
        t1.children.append(t2)
        t1.degree += 1
        return t1

    def insert(self, key):
        new_heap = BinomialHeap()
        new_heap.trees.append(BinomialNode(key))
        self.trees = self.union(new_heap)

    def union(self, other):
        # упрощенная реализация объединения куч по степеням
        pass  # Полная реализация очень объемна.

# пример вставки
bh = BinomialHeap()
bh.insert(10)
bh.insert(20)
    
// Фибоначчиева куча (Fibonacci Heap)
class Node {
    int key;
    int degree;
    Node parent;
    Node child;
    Node left;
    Node right;
    boolean mark;

    public Node(int key) {
        this.key = key;
        this.degree = 0;
        this.parent = null;
        this.child = null;
        this.left = this;
        this.right = this;
        this.mark = false;
    }
}

public class FibonacciHeap {
    private Node min;
    private int n;

    public FibonacciHeap() {
        min = null;
        n = 0;
    }

    // Вставка нового узла
    public void insert(int key) {
        Node node = new Node(key);
        if (min == null) {
            min = node;
        } else {
            // Вставляем в корневой список
            node.left = min;
            node.right = min.right;
            min.right.left = node;
            min.right = node;
            if (key < min.key) {
                min = node;
            }
        }
        n++;
    }

    public Node getMin() {
        return min;
    }

    public int size() {
        return n;
    }

    // Здесь должны быть методы объединения, удаления минимума, уменьшения ключа и т.д.
    
    public static void main(String[] args) {
        FibonacciHeap heap = new FibonacciHeap();
        heap.insert(10);
        heap.insert(4);
        heap.insert(15);
        heap.insert(7);
        System.out.println("Минимальное значение кучи: " + heap.getMin().key);
        System.out.println("Размер кучи: " + heap.size());
    }
}



    // Двоичная куча
    import java.util.PriorityQueue;
import java.util.Collections;
public class BinaryHeap {
    public static void main(String[] args) {
        PriorityQueue<Integer> maxHeap = new PriorityQueue<>(Collections.reverseOrder());
        maxHeap.add(10); maxHeap.add(20); maxHeap.add(15);
        maxHeap.poll();
        for (int i : maxHeap) System.out.print(i + " ");
    }
}


//Hash таблицы 
import java.util.HashMap;

public class HashMapExample {
    public static void main(String[] args) {
        HashMap<String, Integer> map = new HashMap<>();
        map.put("one", 1);
        map.put("two", 2);

        System.out.println(map.get("one"));
    }
}
